beacon_node/beacon_chain/src/validator_monitor.rs

```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                let current_slot = state.slot();
+                self.missed_blocks.contains(&(current_epoch, i as u64, current_slot))
+                    .then(|| {
+                        metrics::inc_counter_vec(
+                            &metrics::VALIDATOR_MONITOR_MISSED_NON_FINALIZED_BLOCKS_TOTAL,
+                            &[current_epoch.to_string().as_str(), current_slot.to_string().as_str(), id],
+                        );
+                    });
+            }
+        }
+    }
+
+    fn add_validators_missed_blocks(&mut self, current_epoch: Epoch, state: &BeaconState<T>) {
+        // Determine missed (non-finalized) blocks (can contain false positives)
+        let range_of_slot = (T::slots_per_epoch() as usize - MISSED_BLOCK_LAG_SLOTS) - 1;
+        for n in 0..range_of_slot {
+            let slot = state.slot() - Slot::new(n as u64);
+            let prev_slot = slot - 1;
+
+            // safeguard if we are on a new chain, the slot - 1 can be negative
+            if prev_slot < 0 {
+                break;
+            }
+
+            // condition for missed_block is defined such as block_root == block_root - n
+            // where the proposer who missed the block is the proposer of the block at block_root - n
+            if let (Ok(block_root), Ok(prev_block_root)) = (state.get_block_root(slot), state.get_block_root(prev_slot)) {
+                if block_root == prev_block_root {
+                    let epoch = slot.epoch(T::slots_per_epoch());
+                    if let Ok(shuffling_decision_block) = state.proposer_shuffling_decision_root(epoch, *block_root) {
+                        // TODO: AI(Joel) Ask the team if we can use the beacon proposer cache without the mutex
+                        // mutex is implemented in the caller (BeaconChain), not in the cache implementation
+                        if let Some(proposer) = self.beacon_proposer_cache.get_slot::<T>(shuffling_decision_block, slot) {
+                            // only add missed blocks for the proposer if it's in the list of monitored validators
+                            if self.validators
+                                .values()
+                                .into_iter()
+                                .find(|validator| validator.index == Some(proposer.index as u64))
+                                .is_some() {
+                                self.missed_blocks.insert((current_epoch, proposer.index as u64, slot));
+                            }
+                        } else {
+                            debug!(
+                                self.log,
+                                "Could not find proposer for a missed non-finalized block in beacon proposer cache";
+                                "slot" => slot,
+                                "block_root" => format!("{}", block_root),
+                                "shuffling_decision_block" => format!("{}", shuffling_decision_block),
```

Comment:
`slog` has a little signifier for using the Display impl:

```suggestion
                                "block_root" => %block_root,
                                "shuffling_decision_block" => %shuffling_decision_block,
```

you can also get Debug, which would show the full block root, with `?name`